### PR TITLE
Evo 1018 port widget data selection

### PR DESCRIPTION
**Pull request by thomasgwatson**
So I want to propose a pattern for us to follow.

As I mention in the readme update: 
_A new pattern being introduced for the application is the addition of `useEnsureX`, which is a custom hook that handles both selecting `X` from the store and, if it isn't present, also handles the fetching of that data._

Take a look at the below initial example `useEnsureCurrentGroup` and tell me what you think about the idea of moving to this idea. In my eyes, the biggest benefit is that when you drop one of these hooks into a component, you know that component won't have an implicit dependency on something else to make the request for the data it needs.

### CODE REVIEW

File: src/hooks/useEnsureCurrentGroup/index.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import fetchGroup from 'store/actions/fetchGroupDetails'
+import { useSelector, useDispatch } from 'react-redux'
+import presentGroup from 'store/presenters/presentGroup'
+import getGroupForCurrentRoute from 'store/selectors/getGroupForCurrentRoute'
+import { useRouter } from 'hooks/useRouter'
+import { createSelector } from 'reselect'
+import isPendingFor from 'store/selectors/isPendingFor'
+import { FETCH_GROUP_DETAILS } from 'store/constants'
+
+const selectAndPresentGroup = createSelector(
```
Comment (thomasgwatson) [inline src/hooks/useEnsureCurrentGroup/index.js:11]: Added an extra reselect here to ensure `presentGroup` wasn't being re-run on every pass through of a component using this hook, and thus possibly forcing unnecessary re-rendering if the result of this hook was passed down to a child component

**Comment by thomasgwatson** [comment]
I had greater ambitions to replace `getWidgetItems` and push those data selections/transformations into their respective widget components but a bunch of the logic around visibility is handled in the `Widget` component. So I don't currently have a good idea about how to handle that...

**Comment by thomasgwatson** [comment]
Closes #1018

File: src/components/Widget/Widget.js [hunk 2]
```diff
@@ -20,6 +20,11 @@ import PrivacyWidget from './PrivacyWidget'
 import RichTextWidget from './RichTextWidget'
 import JoinWidget from './JoinWidget'
 import TopicsWidget from './TopicsWidget'
+import { useSelector, useDispatch } from 'react-redux'
+import { useEnsureCurrentGroup } from 'hooks/useEnsureCurrentGroup'
```
Comment (lorenjohnson) [inline src/components/Widget/Widget.js:24]: Why not default export?

File: src/hooks/useEnsureCurrentGroup/index.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  (state, router) => getGroupForCurrentRoute(state, router),
+  (group) => presentGroup(group)
+)
+
+export function useEnsureCurrentGroup () {
```
Comment (lorenjohnson) [inline src/hooks/useEnsureCurrentGroup/index.js:16]: Why not the default export?

**Review by lorenjohnson — APPROVED** [review]
I don't have the context of having looked at the `farm` branch yet so can't comment on a lot of this. I think the `useEnsureCurrentGroup` hook is great and likely something we'll want to use (sic) more generally once this is merged. As for generalizing on the `useEnsureX` pattern for other areas I'm hesitant because of the things I've named which I've also wadded into just now. Great work though, keep going! And I still think we should take an hour or two and jam/mindmeld/spitball some of this together soon.

**Review by tibetsprague — APPROVED** [review]
cool!

**Pull request merged**